src/platforms/wayland/displayclient.cpp

```diff
@@ -0,0 +1,856 @@
+    if (output)
+        wl_output_destroy(output);
+
+    if (window)
+        wl_shell_surface_destroy(window);
+
+    wl_surface_destroy(surface);
+
+    if (eglsurface != EGL_NO_SURFACE)
+        eglDestroySurface(owner->egldisplay, eglsurface);
+
+    if (eglctx != EGL_NO_CONTEXT)
+        eglDestroyContext(owner->egldisplay, eglctx);
+}
+
+wl_output_listener const mgw::DisplayClient::Output::output_listener = {
+    &geometry,
+    &mode,
+    &done,
+    &scale,
+};
+
+void mgw::DisplayClient::Output::done(void* data, struct wl_output* /*wl_output*/)
+{
+    auto output = static_cast<Output*>(data);
+    output->on_done(*output);
+}
+
+void mgw::DisplayClient::Output::for_each_display_buffer(std::function<void(DisplayBuffer & )> const& f)
+{
+    if (!window)
+    {
+        static wl_shell_surface_listener const shell_surface_listener{
+            [](void*, wl_shell_surface* shell_surface, uint32_t serial){ wl_shell_surface_pong(shell_surface, serial); },
+            [](void*, wl_shell_surface*, uint32_t, int32_t, int32_t){},
+            [](void*, wl_shell_surface*){}
+        };
+
+        window = wl_shell_get_shell_surface(owner->shell, surface);
+        wl_shell_surface_add_listener(window, &shell_surface_listener, this);
+        wl_shell_surface_set_fullscreen(window, WL_SHELL_SURFACE_FULLSCREEN_METHOD_SCALE, 0, output);
+        wl_display_dispatch(owner->display);
+
+        auto const& size = dcout.modes[dcout.current_mode_index].size;
+
+        eglsurface = eglCreatePlatformWindowSurface(
+            owner->egldisplay,
+            owner->eglconfig,
+            wl_egl_window_create(surface, size.width.as_int(), size.height.as_int()), nullptr);
+    }
+
+    f(*this);
+}
+
+void mgw::DisplayClient::Output::post()
+{
+}
+
+auto mgw::DisplayClient::Output::recommended_sleep() const -> std::chrono::milliseconds
+{
+    return std::chrono::milliseconds{0};
+}
+
+auto mgw::DisplayClient::Output::view_area() const -> geometry::Rectangle
+{
+    return dcout.extents();
+}
+
+bool mgw::DisplayClient::Output::overlay(mir::graphics::RenderableList const&)
+{
+    return false;
+}
+
+auto mgw::DisplayClient::Output::transformation() const -> glm::mat2
+{
+    return glm::mat2{1};
+}
+
+auto mgw::DisplayClient::Output::native_display_buffer() -> NativeDisplayBuffer*
+{
+    return this;
+}
+
+void mgw::DisplayClient::Output::make_current()
+{
+    if (eglctx == EGL_NO_CONTEXT)
+        eglctx = eglCreateContext(owner->egldisplay, owner->eglconfig, owner->eglctx, ctxattribs);
+
+    if (!eglMakeCurrent(owner->egldisplay, eglsurface, eglsurface, eglctx))
+        throw std::runtime_error("Can't eglMakeCurrent");
+}
+
+void mgw::DisplayClient::Output::release_current()
+{
+    eglMakeCurrent(owner->egldisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+}
+
+void mgw::DisplayClient::Output::swap_buffers()
+{
+    if (eglSwapBuffers(owner->egldisplay, eglsurface) != EGL_TRUE)
+        fatal_error("Failed to perform buffer swap");
```

Comment:
There doesn't seem to be a particular reason not to be throwing an `mg::egl_error` here, either?